File: infrastructures/ddd/python/src/infrastructure/repositories/infra/i_unit_of_work.py
"""
Interface for Unit of Work pattern.
UnitOfWork manages transactions across multiple repositories.
Transaction management is handled by UnitOfWork middleware automatically.
"""

from abc import ABC, abstractmethod


class IUnitOfWork(ABC):
    """
    Interface for Unit of Work pattern.
    
    UnitOfWork:
    - Manages database transactions
    - Coordinates multiple repository operations
    - Ensures atomicity of business operations
    - Called by UnitOfWorkMiddleware automatically
    - Services MUST NOT manually manage transactions
    """
    
    @abstractmethod
    async def begin_transaction_async(self) -> None:
        """Begins a new database transaction."""
        pass
    
    @abstractmethod
    async def commit_async(self) -> None:
        """Commits the current transaction."""
        pass
    
    @abstractmethod
    async def rollback_async(self) -> None:
        """Rolls back the current transaction."""
        pass
    
    @abstractmethod
    async def save_changes_async(self) -> None:
        """Saves all pending changes to the database."""
        pass

File: infrastructures/ddd/typescript/src/infrastructure/repositories/infra/IUnitOfWork.ts
/**
 * Unit of Work interface for managing database transactions.
 * UnitOfWork:
 * - Manages database transaction lifecycle
 * - Applied automatically by UnitOfWorkMiddleware
 * - Services should NEVER manually manage transactions
 * - Commits on success (status < 400)
 * - Rolls back on error (status >= 400 or exception)
 * 
 * Middleware order (automatically applied):
 * 1. Correlation ID → 2. Logging → 3. Authentication → 
 * 4. Authorization/Ownership → 5. UnitOfWork → 6. Controller →
 * 7. UnitOfWork commit/rollback → 8. Error Handling
 */
export interface IUnitOfWork {
  /**
   * Begins a new database transaction.
   * Called automatically by middleware.
   */
  BeginTransactionAsync(): Promise<void>;

  /**
   * Commits the current transaction.
   * Called automatically by middleware on success.
   */
  CommitAsync(): Promise<void>;

  /**
   * Rolls back the current transaction.
   * Called automatically by middleware on error.
   */
  RollbackAsync(): Promise<void>;

  /**
   * Saves all changes to the database within the current transaction.
   * Optional - depends on ORM implementation.
   */
  SaveChangesAsync(): Promise<void>;
}
